File: 206-reverse-linked-list/206-reverse-linked-list.py
class Solution:
    def reverse(self, prev, node):
        if not  node:
            return prev
        
        temp = node.next
        node.next = prev
        return self.reverse(node, temp)
    
    def reverseList(self, node: Optional[ListNode]) -> Optional[ListNode]:
        return self.reverse(None, node)

File: 206-reverse-linked-list/206-reverse-linked-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseList(head *ListNode) *ListNode {
    if head == nil || head.Next == nil {
        return head
    }
    
    one, two, three := head, head.Next, head.Next.Next    
    one.Next = nil
    
    for two != nil {
        two.Next = one
        one = two
        two = three
        
        if three != nil {
            three = three.Next                
        }
    }
    
    return one
}